### ISSUE TITLE
No repositories after restart and issues creating repository on existing S3 bucket

### ISSUE THREAD

**Issue by FredrikBakken** [treeverse/lakeFS#6256]
### What happened?

**Current Behavior:**
Creating repository on already existing S3 storage results in the following error message:
```sh
failed to create repository: found lakeFS objects in the storage namespace(s3://test): storage namespace already in use
```
        
**Steps to Reproduce:**
1. Create a `docker-compose.yml` file with the following contents:
   ```
   version: '3.9'
   
   services:
     minio:
       image: minio/minio:RELEASE.2023-06-29T05-12-28Z
       container_name: minio
       ports:
         - "9000:9000"
         - "9001:9001"
       entrypoint: ["minio", "server", "/data", "--console-address", ":9001"]
  
     postgres:
       image: postgres:14
       container_name: postgres
       environment:
         POSTGRES_USER: postgres
         POSTGRES_PASSWORD: postgres
       healthcheck:
         test: ["CMD", "pg_isready", "-U", "postgres"]
         interval: 10s
         retries: 5
         start_period: 5s
       restart: always

     minio-setup:
       image: minio/mc:RELEASE.2023-06-28T21-54-17Z
       container_name: minio-setup
       environment:
           - MC_HOST_lakefs=http://minioadmin:minioadmin@minio:9000
       depends_on:
         - minio
       entrypoint: >
         /bin/sh -c "
         /usr/bin/mc mb lakefs/test;
         exit 0;
         "

     lakefs:
       image: treeverse/lakefs:0.104.0
       container_name: lakefs
       depends_on:
         - postgres
         - minio-setup
       ports:
         - "8000:8000"
       environment:
         - LAKEFS_DATABASE_TYPE=local
         - LAKEFS_BLOCKSTORE_TYPE=s3
         - LAKEFS_BLOCKSTORE_S3_FORCE_PATH_STYLE=true
         - LAKEFS_BLOCKSTORE_S3_ENDPOINT=http://minio:9000
         - LAKEFS_BLOCKSTORE_S3_CREDENTIALS_ACCESS_KEY_ID=minioadmin
         - LAKEFS_BLOCKSTORE_S3_CREDENTIALS_SECRET_ACCESS_KEY=minioadmin
         - LAKEFS_AUTH_ENCRYPT_SECRET_KEY=some random secret string
         - LAKEFS_STATS_ENABLED
         - LAKEFS_LOGGING_LEVEL
         - LAKECTL_CREDENTIALS_ACCESS_KEY_ID=AKIAIOSFODNN7EXAMPLE
         - LAKECTL_CREDENTIALS_SECRET_ACCESS_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
         - LAKECTL_SERVER_ENDPOINT_URL=http://localhost:8000
         - LAKEFS_DATABASE_CONNECTION_STRING=postgres://postgres:postgres@postgres-lakefs/postgres?sslmode=disable
       entrypoint: ["/bin/sh", "-c"]
       command:
           - |
             lakefs setup --local-settings --user-name docker --access-key-id AKIAIOSFODNN7EXAMPLE --secret-access-key wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY || true
             lakefs run --local-settings &
             wait-for -t 60 lakefs:8000 -- lakectl repo create lakefs://test s3://test || true
             wait
   ```
3. Start this docker container environment by running: `docker compose up`
4. Go to the `test` repository in the UI on http://localhost:8000 and upload a file, e.g. `test.txt`
5. Simulate a restart of LakeFS by running the following two commands in the terminal:
   ```sh
   docker compose down lakefs
   docker compose up
   ```
7. Now open the LakeFS UI again, where you now can see that no repositories are found
8. Try to re-create repository in the UI by creating an empty repository with the following details:
   - **Repository ID**: test
   - **Storage Namespace**: s3://test
   - **Default Branch**: main

### Expected Behavior

**Expected Behavior:**

Expecting behavior is that after a restart of LakeFS, the repositories are still visible in the UI. If they do not show up, make it possible to create repository on existing S3 storage.

Using the import functionality does not seem to satisfy the requirements as the files are of binary format that does not have the same naming scheme as previously uploaded.

### lakeFS Version

0.104.0

### Deplyoment

Local with min.io, and Azure Container Instance with blob storage

**Comment by rmoff** [comment]
Hi @FredrikBakken! 

The behaviour that you're observing is exactly as designed. lakeFS stores its metadata on object store (minio in this case), and restarting the lakeFS server does not remove that. 

If you want to reset your environment then `docker compose down && docker compose up` will do that (i.e. recreate the minio container to delete its objects)

It's a nice idea lakeFS actually scanning the attached storage for repositories and pre-populating them in the UI, but that would be new functionality for us to build. 

Can you elaborate a bit more on what your requirements/use-case is here please? I didn't quite follow your comment about the use of import and how it relates to the repository already existing in the object store.

**Comment by FredrikBakken** [comment]
Hi @rmoff ! 

Thank you for the quick follow-up to my Issue.

The problem I've encountered is related to LakeFS not being able to re-populate any repositories in the UI after the container is stopped (on purpose or after crashing) and then started again. In my case I have a self-hosted LakeFS container mapped to S3 buckets in min.io, once the LakeFS container is stopped and started the repositories defined in LakeFS before the restart does no longer appear in the UI (no repositories are listed in LakeFS). However, all the data and metadata objects can still be found inside the min.io buckets as I in this example only stopped and started LakeFS.

When I then try to re-create the repository in the LakeFS UI on the existing min.io buckets with data and metadata, I am getting the error message:

```sh
failed to create repository: found lakeFS objects in the storage namespace(s3://test): storage namespace already in use
```

The behaviour I'd want to see is that I can re-map the repositories to the S3 buckets in the restarted LakeFS container.

I hope this description is a bit easier to understand. Please let me know if some details still are unclear and I can create an even more detailed step-by-step example.

**Comment by FredrikBakken** [comment]
@rmoff , you've absolutely made my day with this explanation - **thank you!** 🚀

I now clearly understand where my misconfigurations came into causing chaos for my local set up. It also helped me understand why I was seeing a similar issues in Azure where environment variables were actually set up correctly, but since it was spun up together with PostgreSQL as a Terraform module the metadata was reset when it was taken down and spun up, while Storage Account with Container and data was not taken down and spun up with the rest - adding this as it might help others in Azure.

When it comes to the: `docker compose down lakefs`, I am not sure why it does not work for you. I have my terminal placed in the same directory as where the `docker-compose.yml` file exists. Maybe it is related to `docker` / `docker compose` versions? Mine are:

```
$ docker version
Client: Docker Engine - Community
 Version:           24.0.5
 API version:       1.43
 Go version:        go1.20.6
 Git commit:        ced0996
 Built:             Fri Jul 21 20:35:18 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          24.0.4
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.5
  Git commit:       4ffc614
  Built:            Fri Jul  7 14:50:55 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.21
  GitCommit:        3dce8eb055cbb6872793272b4f20ed16117344f8
 runc:
  Version:          1.1.7
  GitCommit:        v1.1.7-0-g860f061
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0

$ docker compose version
Docker Compose version v2.20.2
```

**Comment by FredrikBakken** [comment]
Another addition that I came over during my testing. I've encountered race conditions when trying to start the entire container environment with `docker compose up`, where I had some issues with trying to add the [automated wait](https://github.com/treeverse/lakeFS/blob/f5bf238a6be454af2d98bd8327e52ed9443c63b0/esti/ops/docker-compose.yaml#L31C12-L31C12).

Instead I changed the `depends_on` to the following:

```sh
depends_on:
  postgres:
    condition: service_healthy
  minio-setup:
    condition: service_completed_successfully
```

Which seems to be a docker native solution to avoiding race conditions in this case.
